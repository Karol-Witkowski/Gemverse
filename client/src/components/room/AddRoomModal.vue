<template>
  <v-card cols="12">
    <v-card-title class="headline grey--text text--darken-2">
      Add new room
    </v-card-title>
    <v-card-text>
      <v-form
        ref="form"
        v-model="isFormValid"
      >
        <v-col cols="12">
          <v-text-field
            autofocus
            :counter="15"
            :error-messages="error"
            hint="Required"
            id="name"
            label="Room Name"
            persistent-hint
            required
            :rules="nameRules"
            v-model.trim="room.name"
            v-on:keyup="error = ''"
            v-on:keyup.enter="formValidation"
          />
        </v-col>
        <v-col cols="12">
          <v-text-field
            label="Password - optional"
            :rules="passwordRules"
            type="password"
            v-model.trim="room.password"
            v-on:keyup.enter="formValidation"
          />
        </v-col>
      </v-form>
    </v-card-text>
    <v-card-actions class="pb-4">
      <v-btn
        @click="closeModal"
        color="primary"
        text
        outlined
      >
        Close
      </v-btn>
      <v-spacer />
      <v-btn
        @click.prevent="formValidation"
        color="primary"
        :disabled="!isFormValid"
        text
        type="submit"
        outlined
      >
        Save
      </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script>
import axios from 'axios';
import { mapGetters } from 'vuex';
import * as io from 'socket.io-client';

export default {
  name: 'AddRoomModal',
  data() {
    return {
      error: [],
      isFormValid: false,
      nameRules: [
        (value) => (value.length >= 3 && value.length <= 15) || 'Characters range: 3 - 15',
        (value) => !!value || 'Required.',
      ],
      passwordRules: [
        (value) => !(/[ ]/.test(value)) || 'No blank spaces allowed',
        (value) => ((value.length === 0 || value.length >= 6) && value.length <= 128) || 'Password must be at least 6 characters long',
      ],
      room: {
        name: '',
        password: '',
        user: {},
      },
      socket: io('http://localhost:3000'),
    };
  },

  computed: {
    ...mapGetters(['getUserInfo']),
  },

  methods: {
    closeModal() {
      this.$emit('close-modal');
      this.error = '';
      this.resetData();
      this.$refs.form.resetValidation();
    },

    createRoom() {
      axios.post('http://localhost:3000/api/room', this.room)
        .then((response) => {
          this.socket.emit('createRoom', this.room.name, this.room.password, response.data.slug);
          this.resetData();
          if (response.status === 201) {
            this.closeModal();
          }
        })
        .catch((error) => {
          console.log(error);
          this.error = error.response.data.error;
        });
    },

    formValidation() {
      if (this.isFormValid) {
        this.room.user = this.getUserInfo;
        this.createRoom();
      }
    },

    resetData() {
      this.room.name = '';
      this.room.password = '';
    },
  },
};
</script>
