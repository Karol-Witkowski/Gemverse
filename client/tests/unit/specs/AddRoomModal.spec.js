import { createLocalVue, mount } from "@vue/test-utils";
import axios from "axios";
import io from "socket.io-client";
import MockedSocket from "socket.io-mock";
import Vue from "vue";
import Vuetify from "vuetify";
import AddRoomModal from "@/components/room/AddRoomModal.vue";

const localVue = createLocalVue();
let socket;
const url = "http://localhost:3000/api/room";
const vuetify = new Vuetify();
let wrapper;
const mockStore = {
  getters: {
    getUserInfo: {
      createdDate: "2021-04-20T01:01:22.269Z",
      email: "test@mail.js",
      _id: "321testid",
      username: "testUser",
    },
  },
};
const error = {
  response: {
    data: {
      errors: {
        name: {
          msg: "Name error",
        },
        password: {
          msg: "Password error",
        },
      },
    },
  },
};
const response = {
  data: {
    data: "testMessageData",
    success: true,
  },
};

jest.mock("axios");
jest.mock("socket.io-client");

afterEach(() => {
  jest.resetAllMocks();
  wrapper.destroy();
});

describe("Implementation test for AddRoomModal.vue - successful HTTP post", () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    socket = new MockedSocket();
    io.mockReturnValue(socket);

    wrapper = mount(AddRoomModal, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          isFormValid: false,
          room: {
            name: "",
            password: "",
          },
        };
      },
    });
  });

  it("Render correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Check that socket connection is established", () => {
    expect(io.connect).toHaveBeenCalled();
  });

  it("Should emit new room data", () => {
    socket.on("createRoom", (data) => {
      expect(data).toEqual("roomData");
    });
  });

  it("Initializes with correct elements", () => {
    // Check buttons initial state
    expect(wrapper.findAll("button").length).toEqual(2);
    expect(wrapper.find("button[name=close]").text()).toMatch("close");
    expect(wrapper.find("button[name=save]").text()).toMatch("save");
    expect(wrapper.find("button[name=save]").element.disabled).toBeTruthy();

    // Check inputs initial state
    expect(wrapper.findAll(".v-text-field").length).toEqual(2);
    expect(wrapper.find("input[name=name]").text()).toEqual("");
    expect(wrapper.find("input[name=name]").text()).toEqual("");

    // Check validation initial state
    expect(wrapper.findAll(".v-messages").length).toEqual(2);
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");
    expect(wrapper.vm.isformValid).toBeFalsy();
  });

  it("Set input data correctly", async () => {
    // Check inputs initial state
    expect(wrapper.findAll(".v-text-field").length).toEqual(2);
    expect(wrapper.find("input[name=name]").text()).toEqual("");
    expect(wrapper.find("input[name=password]").text()).toEqual("");

    await wrapper.setData({
      room: {
        name: "test",
        password: "123456",
      },
    });

    await Vue.nextTick();

    // Check that the room data was properly set
    expect(wrapper.find("input[name=name]").element.value).toEqual("test");
    expect(wrapper.find("input[name=password").element.value).toEqual("123456");
  });

  it("Pass validation when the only name was entered", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
      },
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeTruthy();

    // Check that correct input messages are displayed
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");
  });

  it("Fail validation when room name character range is invalid", async () => {
    await wrapper.setData({
      room: {
        name: "1",
      },
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeFalsy();

    // Check that correct input messages are displayed
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Characters range: 3 - 15");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");
  });

  it("Fail validation when the only password was entered", async () => {
    await wrapper.setData({
      room: {
        password: "123456",
      },
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeFalsy();

    // Check that correct input messages are displayed
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");
  });

  it("Fail validation when password contains whitespace", async () => {
    await wrapper.setData({
      room: {
        name: "test",
        password: "123 456",
      },
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeFalsy();

    // Check that correct input messages are displayed
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("No blank spaces allowed");
  });

  it("Fail validation when room password character range is invalid", async () => {
    await wrapper.setData({
      room: {
        name: "test",
        password: "123",
      },
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeFalsy();

    // Check that correct input messages are displayed
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual(
      "Password must be at least 6 characters long"
    );
  });

  it("Enables save room button on validation pass", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
        password: "123456",
      },
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeTruthy();

    // Check that correct input messages are displayed
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");

    // Check that the save button is active
    expect(wrapper.find("button[name=save]").element.disabled).toBeFalsy();
  });

  it("Should send a post request with correct data on submit", async () => {
    await wrapper.setData({
      room: {
        name: "name",
        password: "123456",
      },
    });

    await Vue.nextTick();

    wrapper.vm.formValidation();

    // Check if a post was called
    expect(axios.post).toHaveBeenCalled();

    // Check if a post was called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if a post was called with the correct data
    expect(axios.post).toHaveBeenCalledWith(url, {
      name: "name",
      password: "123456",
      user: {
        _id: "321testid",
        createdDate: "2021-04-20T01:01:22.269Z",
        email: "test@mail.js",
        username: "testUser",
      },
    });
  });

  it("Should emit close dialog event", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
        password: "123456",
      },
    });

    await wrapper.vm.createRoom();

    expect(wrapper.emitted().closeModal).toBeTruthy();
    expect(wrapper.emitted().closeModal.length).toBe(1);
  });
});

beforeEach(() => {
  wrapper = mount(AddRoomModal, {
    localVue,
    mocks: {
      $store: mockStore,
    },
    vuetify,
    data() {
      return {
        isFormValid: false,
        nameError: "",
        passwordError: "",
        room: {
          name: "",
          password: "",
        },
      };
    },
  });
});

describe("Implementation test for AddRoomModal.vue - failed HTTP post", () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(error);
  });

  it("Should display errors", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
        password: "123456",
      },
    });

    await wrapper.vm.createRoom();

    await Vue.nextTick();

    expect(wrapper.findAll(".v-messages").length).toEqual(2);
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Name error");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("Password error");
    expect(wrapper.vm.isformValid).toBeFalsy();
  });
});

describe("Behavioral test for AddRoomModal.vue - successful HTTP post", () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);
  });

  it("Should not send post request when inputs are empty", async () => {
    await wrapper.find("button[name=save]").trigger("click");

    expect(axios.post).not.toHaveBeenCalled();
  });

  it("Set input data correctly", async () => {
    await wrapper.find("input[name=name]").setValue("test");
    await wrapper.find("input[name=password]").setValue("123456");

    await Vue.nextTick();

    // Check that the room data was properly set
    expect(wrapper.find("input[name=name]").element.value).toEqual("test");
    expect(wrapper.find("input[name=password]").element.value).toEqual("123456");
  });

  it("Should send a post request with correct data on submit", async () => {
    await wrapper.find("input[name=name]").setValue("test");
    await wrapper.find("input[name=password]").setValue("123456");

    await Vue.nextTick();

    wrapper.find("button[name=save]").trigger("click");

    // Check if a post was called
    expect(axios.post).toHaveBeenCalled();

    // Check if a post was called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if a post was called with the correct data
    expect(axios.post).toHaveBeenCalledWith(url, {
      name: "test",
      password: "123456",
      user: {
        _id: "321testid",
        createdDate: "2021-04-20T01:01:22.269Z",
        email: "test@mail.js",
        username: "testUser",
      },
    });
  });

  it("Should emit close dialog event", async () => {
    await wrapper.find("input[name=name]").setValue("test");
    await wrapper.find("input[name=password").setValue("123456");

    await Vue.nextTick();

    wrapper.findAll("button[name=save]").trigger("click");

    expect(wrapper.emitted()).toBeTruthy();
  });
});

describe("Behavioral test for AddRoomModal.vue - failed HTTP post", () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(error);
  });

  it("Should display errors", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
        password: "123456",
      },
    });

    await Vue.nextTick();

    await wrapper.find("button[name=save]").trigger("click");

    await Vue.nextTick();

    expect(wrapper.findAll(".v-messages").length).toEqual(2);
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Name error");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("Password error");
    expect(wrapper.vm.isformValid).toBeFalsy();
  });

  it("Should reset errors on keyup event", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
        password: "123456",
      },
    });

    await Vue.nextTick();

    await wrapper.findAll("button[name=save]").trigger("click");

    await Vue.nextTick();

    // Check if errors are visible
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Name error");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("Password error");

    await wrapper.findAll("input").trigger("keyup");

    // Check if keyup event restored messages
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");
  });

  it("Should reset errors on close event", async () => {
    await wrapper.setData({
      room: {
        name: "room name",
        password: "123456",
      },
    });

    await Vue.nextTick();

    await wrapper.find("button[name=save]").trigger("click");

    await Vue.nextTick();

    // Check if errors are visible
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Name error");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("Password error");

    await wrapper.find("button[name=close]").trigger("click");

    // Check if closing restored messages
    expect(wrapper.findAll(".v-messages").at(0).text()).toEqual("Required");
    expect(wrapper.findAll(".v-messages").at(1).text()).toEqual("");
  });
});
