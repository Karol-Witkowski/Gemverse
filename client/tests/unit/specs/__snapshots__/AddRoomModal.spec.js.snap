// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for AddRoomModal.vue - successful HTTP post Render correctly 1`] = `
<div class="v-card v-sheet theme--light" cols="12">
  <div class="v-card__title headline grey--text text--darken-2">
    Add new room
  </div>
  <div class="v-card__text">
    <form novalidate="novalidate" class="v-form">
      <div class="col col-12">
        <div class="v-input theme--light v-text-field">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="name" class="v-label theme--light" style="left: 0px; position: absolute;">Room Name</label><input name="name" required="required" autofocus="autofocus" id="name" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                  <div class="v-messages__message">Required</div>
                </transition-group-stub>
              </div>
              <div class="v-counter theme--light">0 / 15</div>
            </div>
          </div>
        </div>
      </div>
      <div class="col col-12">
        <div class="v-input theme--light v-text-field">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-7" class="v-label theme--light" style="left: 0px; position: absolute;">Password - optional</label><input name="password" id="input-7" type="password"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div class="v-card__actions pb-4"><button type="button" class="v-btn v-btn--outlined v-btn--text theme--light v-size--default primary--text" name="close"><span class="v-btn__content">
      close
    </span></button>
    <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--outlined v-btn--text theme--light v-size--default" name="save"><span class="v-btn__content">
      save
    </span></button>
  </div>
</div>
`;
